### PR TITLE
[NET-5346] Expose JWKCluster fields in jwt-provider config entry

### PR DESCRIPTION
**Pull request by roncodingenthusiast**
Changes proposed in this PR:
- After exposing JWKCluster https://github.com/hashicorp/consul/pull/17978 we want to ensure the CRD for jwt-provider config entry matches the one in consul core

How I've tested this PR:
- Manually
- Unit tests

How I expect reviewers to test this PR:

TODO:
- [x] Tests added
- [x] changelog entry

### CODE REVIEW

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 1]
```diff
@@ -432,6 +441,180 @@ func (r *RemoteJWKS) validate(path *field.Path) field.ErrorList {
...
+	// TLSCertificates refers to the data containing certificate authority certificates to use
+	// in verifying a presented peer certificate.
+	// If not specified and a peer certificate is presented it will not be verified.
+	//
+	// Must be either CaCertificateProviderInstance or TrustedCA.
+	TLSCertificates *JWKSTLSCertificate `json:",omitempty" alias:"tls_certificates"`
+
+	// The timeout for new network connections to hosts in the cluster.
+	// If not set, a default value of 5s will be used.
+	ConnectTimeout time.Duration `json:",omitempty" alias:"connect_timeout"`
```
Comment (kisunji) [inline control-plane/api/v1alpha1/jwtprovider_types.go:464]: I think the json tags here should have the camelCase alias and not the `alias` tags

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 2]
```diff
@@ -432,6 +441,180 @@ func (r *RemoteJWKS) validate(path *field.Path) field.ErrorList {
...
+// Must be either CaCertificateProviderInstance or TrustedCA.
+type JWKSTLSCertificate struct {
+	// CaCertificateProviderInstance Certificate provider instance for fetching TLS certificates.
+	CaCertificateProviderInstance *JWKSTLSCertProviderInstance `json:",omitempty" alias:"ca_certificate_provider_instance"`
+
+	// TrustedCA defines TLS certificate data containing certificate authority certificates
+	// to use in verifying a presented peer certificate.
+	//
+	// Exactly one of Filename, EnvironmentVariable, InlineString or InlineBytes must be specified.
+	TrustedCA *JWKSTLSCertTrustedCA `json:",omitempty" alias:"trusted_ca"`
```
Comment (kisunji) [inline control-plane/api/v1alpha1/jwtprovider_types.go:521]: Same json tag comments as above

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 3]
```diff
@@ -432,6 +441,180 @@ func (r *RemoteJWKS) validate(path *field.Path) field.ErrorList {
...
+	//
+	// The default value is "default".
+	InstanceName string `json:",omitempty" alias:"instance_name"`
+
+	// CertificateName is used to specify certificate instances or types. For example, "ROOTCA" to specify
+	// a root-certificate (validation context) or "example.com" to specify a certificate for a
+	// particular domain.
+	//
+	// The default value is the empty string.
+	CertificateName string `json:",omitempty" alias:"certificate_name"`
```
Comment (kisunji) [inline control-plane/api/v1alpha1/jwtprovider_types.go:565]: same json tag comments as above

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 4]
```diff
@@ -432,6 +441,180 @@ func (r *RemoteJWKS) validate(path *field.Path) field.ErrorList {
...
+
+// JWKSTLSCertTrustedCA defines TLS certificate data containing certificate authority certificates
+// to use in verifying a presented peer certificate.
+//
+// Exactly one of Filename, EnvironmentVariable, InlineString or InlineBytes must be specified.
+type JWKSTLSCertTrustedCA struct {
+	Filename            string `json:",omitempty" alias:"filename"`
+	EnvironmentVariable string `json:",omitempty" alias:"environment_variable"`
+	InlineString        string `json:",omitempty" alias:"inline_string"`
+	InlineBytes         []byte `json:",omitempty" alias:"inline_bytes"`
```
Comment (kisunji) [inline control-plane/api/v1alpha1/jwtprovider_types.go:587]: same json tag comments as above

**Review by kisunji — COMMENTED** [review]
Just one recurring comment about json tags but the rest look good.
I think you may be missing an update to the deepcopy function (I had one in my PR but forgot what generated it 🤔 )

**Comment by roncodingenthusiast** [comment]
Working on adding the acceptance tests, should be ready soon

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 5]
```diff
@@ -432,6 +441,180 @@ func (r *RemoteJWKS) validate(path *field.Path) field.ErrorList {
...
+
+	errs = append(errs, c.DiscoveryType.Validate(path.Child("discoveryType"))...)
+	errs = append(errs, c.TLSCertificates.validate(path.Child("tlsCertificates"))...)
+
+	return errs
+}
+
+type ClusterDiscoveryType string
+
+func (d ClusterDiscoveryType) Validate(path *field.Path) field.ErrorList {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/jwtprovider_types.go:492]: add a comment for a public method

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 6]
```diff
@@ -432,6 +441,180 @@ func (r *RemoteJWKS) validate(path *field.Path) field.ErrorList {
...
+		asJSON, _ := json.Marshal(c)
+		errs = append(errs, field.Invalid(path, string(asJSON), "exactly one of 'trustedCa' or 'caCertificateProviderInstance' is required"))
+	}
+
+	errs = append(errs, c.TrustedCA.validate(path.Child("trustedCa"))...)
+
+	return errs
+}
+
+type JWKSTLSCertProviderInstance struct {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/jwtprovider_types.go:554]: add a comment for a public struct

**Review by wilkermichael — APPROVED** [review]
Great work here! Some really minor comments

Thanks for testing this locally as well :D

File: control-plane/api/v1alpha1/jwtprovider_types.go [hunk 7]
```diff
@@ -432,6 +441,180 @@ func (r *RemoteJWKS) validate(path *field.Path) field.ErrorList {
 	}
 
 	errs = append(errs, r.RetryPolicy.validate(path.Child("retryPolicy"))...)
+	errs = append(errs, r.JWKSCluster.validate(path.Child("jwksCluster"))...)
+	return errs
+}
+
+type JWKSCluster struct {
```
Comment (wilkermichael) [inline control-plane/api/v1alpha1/jwtprovider_types.go:448]: Missed this one, add a comment here too

**Review by kisunji — APPROVED** [review]

**Pull request merged**